veps/sig-storage/incremental-backup.md

```diff
@@ -0,0 +1,330 @@
+In push mode, the data pushed to the PVC will include only the data changes up to the backup point. In pull mode, the user will be asked to request only the dirty bitmap blocks.
+
+A challenge in Kubevirt is that Libvirt is recreated each time a VM restarts, causing any existing checkpoints to be lost. To continue incremental backups after a restart, the checkpoint metadata must be provided to libvirt during VM initialization. This allows libvirt to redefine the checkpoints, ensuring the incremental backup process can resume from the correct point in time.
+
+Two approaches are being considered to be able to redefine the checkpoints after a VM restart:
+- Create a `Checkpoint` CR to store the XML data, either as a plain string or base64 encoded. This approach is more visible and easier to query using the k8s API, but could result in a large number of checkpoints per VM, potentially causing performance issues due to the increased number of objects in the cluster.
+- Store checkpoint metadata alongside the QCOW2 overlay in the VM state PVC to ensure persistence. The checkpoint will be stored as an XML, which takes up very little space. This method does not introduce another CR, but since it is not a Kubernetes resource, it requires an alternative method to get checkpoints information. To access the existing checkpoints data, it will be possible to either execute a virsh command through the virt-launcher pod or a new API will need to be created to provide this info. Alternatively, or additionally, the backup vendor can use the backup's checkpoint name, which will be updated in the VirtualMachineBackup CR status, to manage the checkpoints list themselves, providing their users with checkpoints information through their API.
+
+### Offline backups
+
+Another challenge that needs to be addressed is taking offline backups. Currently, there is no option to create offline backups in Libvirt and QEMU. Backups require QEMU, as it is necessary to interpret the QCOW2 metadata/bitmap.
+The proposed solution is to start the VM in a paused state to take the backup, and then either unpause or shut it down once the process is complete.
+
+### State interruptions during backup
+
+According to the previous solution, if the VM was offline when the backup was requested, we will start the VM in a paused state. When the VM is in a paused state, there is no issue with starting it during the backup process, as Libvirt and QEMU guarantee that any VM operations, such as overwriting blocks in the data file, will not affect the backup's consistency. This consistency assurance could not have been guaranteed if an alternative, non-built-in solution had been considered.
+
+In the case of an online backup if the user tries to stop the VM there will be a built-in [solution](https://issues.redhat.com/browse/RHEL-8067) that Libvirt is currently working on. Until then, we may need to handle it on the Kubevirt side by preventing changes to the VM state during the backup.
```

Comment:
I guess we could pause the VM and then stop if once the VMBackup resource is deleted.

Reply:
We should address here what happens if a VM crashes.  In this case, the CBT information may be corrupted and must be discarded.  Qemu offers an API to check if the dirty bitmaps are clean.  We should check this at VM startup when repopulating libvirt with the checkpoint information.  If a corrupted bitmap is found then we will discard the checkpoints and the next backup will need to be a full backup.

Reply:
Yeah I mentioned it as one of the implementation phases but sure Ill add a paragraph describing this